test/Microsoft.TemplateEngine.Utils.UnitTests/SemanticVersionTests.cs

```diff
@@ -0,0 +1,207 @@
+using System;
+using System.Collections.Generic;
+using System.Text;
+using Xunit;
+
+namespace Microsoft.TemplateEngine.Utils.UnitTests
+{
+    public class SemanticVersionTests
+    {
+        [Theory(DisplayName = nameof(SemanticVersionParse))]
+        [InlineData("1", true, 1, 0, 0)]
+        [InlineData("1-preview1", true, 1, 0, 0, "preview1")]
+        [InlineData("1+build1", true, 1, 0, 0, null, "build1")]
+        [InlineData("1-preview1+build1", true, 1, 0, 0, "preview1", "build1")]
+        [InlineData("1.2", true, 1, 2, 0)]
+        [InlineData("1.2-preview1", true, 1, 2, 0, "preview1")]
+        [InlineData("1.2+build1", true, 1, 2, 0, null, "build1")]
+        [InlineData("1.2-preview1+build1", true, 1, 2, 0, "preview1", "build1")]
+        [InlineData("1.2.3", true, 1, 2, 3)]
+        [InlineData("1.2.3-preview1", true, 1, 2, 3, "preview1")]
+        [InlineData("1.2.3+build1", true, 1, 2, 3, null, "build1")]
+        [InlineData("1.2.3-preview1+build1", true, 1, 2, 3, "preview1", "build1")]
+        [InlineData("0.0.0", true)]
+        [InlineData("01.0.0", false)]
+        [InlineData("0.01.0", false)]
+        [InlineData("0.0.01", false)]
+        [InlineData("1.0.0-", false)]
+        [InlineData("1.0.0+", false)]
+        [InlineData("1.0.0-+", false)]
+        [InlineData("1.0.0-1.1", true, 1, 0, 0, "1.1")]
+        [InlineData("1.0.0-1!1", false)]
+        [InlineData("1.0.0-00", true, 1, 0, 0, "00")] //This is acutally valid, 00 just won't be treated as a numeric segment
```

Comment:
This one and the one below may change depending on whether the check to ensure any digit-only sections have leading zeros gets added to TryParse